activesupport/lib/active_support/duration/iso8601_parser.rb

```diff
@@ -0,0 +1,72 @@
+require 'active_support/core_ext/hash/slice'
+
+module ActiveSupport
+  class Duration
+    # Parses a string formatted according to ISO 8601 Duration into the hash .
+    #
+    # See http://en.wikipedia.org/wiki/ISO_8601#Durations
+    # Parts of code are taken from ISO8601 gem by Arnau Siches (@arnau).
+    # This parser isn't so strict and allows negative parts to be present in pattern.
+    class ISO8601Parser
+      attr_reader :parts
+
+      class ParsingError < ::StandardError; end
+
+      def initialize(iso8601duration)
+        @raw = iso8601duration
+        parse_iso_duration!
+        construct_parts
+        validate_parts!
+      end
+
+      private
+        def parse_iso_duration!
+          @match = @raw.match(/^
+                      (?<sign>\+|-)?
```

Comment:
I haven't seen any and don't want to use such scary things, but it's got covered as well. Someone definitely will need it someday.
```
> d = ActiveSupport::Duration.parse('P-1Y-2M3DT-4H-5M-6S')
=> -1 years, -2 months, 3 days, -4 hours, -5 minutes, and -6 seconds
> now = Time.current
=> Tue, 03 Nov 2015 19:54:50 MSK +03:00
> now + d
=> Sat, 06 Sep 2014 15:49:44 MSK +04:00
> now - 1.year - 2.month + 3.days - 4.hours - 5.minutes - 6.seconds
=> Sat, 06 Sep 2014 15:49:44 MSK +04:00
```